web/frontend/src/pages/ballot/Show.tsx

```diff
@@ -80,76 +92,90 @@ const Ballot: FC = () => {
   };
 
   const handleClick = () => {
-    if (choice === '') {
-      userErrors.noCandidate = t('noCandidate');
-      setUserErrors(userErrors);
-      return;
+    if (ballotIsValid(sortedQuestions, answers, setAnswers)) {
+      setUserErrors('');
+      sendBallot();
+    } else {
+      setUserErrors(t('incompleteBallot'));
     }
-    sendBallot();
-    userErrors.noCandidate = '';
-    setUserErrors(userErrors);
   };
 
-  const electionClosedDisplay = () => {
-    return <div> {t('voteImpossible')}</div>;
-  };
+  const subjectTree = (sorted: Question[], parentId: ID) => {
+    const questions = sorted.filter((question) => question.ParentID === parentId);
```

Comment:
That's pretty inefficient. Why do you need an ordered "flat" list of elements after all ? 
Each subject should have: 

1. The list of element IDs, to specify the subject elements' order
2. A dictionary that maps an element ID to the element (Rank, Select, Subject, Text).

Like so you can do: 

```ts
const SubjectTree = (subject) => {
for id in subject.Order {
  const element  = subject.elements[id];
  {element.Type === SUBJECT ? <SubjectTree subject={subject}/>}
  {element.Type === RANK ? <Rank rank={element}}
  ...
}
```

Then all you need to do is to construct such Array of subjects, that contains the Order, and the Map (replaces what is done in `useConfiguration`. Also, instead of having `question.render...` it would be nicer to directly insert the component like `<Rank/>`.

Reply:
This indeed looks a looot better ! I wasn't sure at first how to recursively display elements, so I fell across this tutorial where they flattened their data, hence the mess in useConfiguration .... 😅